Source/Csla/Server/DataPortal.cs

```diff
@@ -29,63 +30,44 @@ public class DataPortal : IDataPortalServer
 
     /// <summary>
     /// Creates an instance of the type.
     /// </summary>
     /// <param name="applicationContext"></param>
     /// <param name="dashboard"></param>
-    public DataPortal(ApplicationContext applicationContext, IDashboard dashboard)
+    /// <param name="options"></param>
+    /// <param name="activator"></param>
+    /// <param name="authorizer"></param>
+    /// <param name="exceptionInspector"></param>
+    /// <param name="factoryLoader"></param>
+    /// <param name="interceptors"></param>
+    public DataPortal(
+      ApplicationContext applicationContext, 
+      IDashboard dashboard, 
+      DataPortalServerOptions options,
+      IAuthorizeDataPortal authorizer,
+      List<IInterceptDataPortal> interceptors,
+      IObjectFactoryLoader factoryLoader,
+      IDataPortalActivator activator,
+      IDataPortalExceptionInspector exceptionInspector)
     {
       ApplicationContext = applicationContext;
       Dashboard = dashboard;
-      var authProviderType = GetAuthProviderType("CslaAuthorizationProvider");
-      if (null == authProviderType)
-        throw new ArgumentNullException(nameof(authProviderType), Resources.CslaAuthenticationProviderNotSet);
-      if (!typeof(IAuthorizeDataPortal).IsAssignableFrom(authProviderType))
-        throw new ArgumentException(Resources.AuthenticationProviderDoesNotImplementIAuthorizeDataPortal, nameof(authProviderType));
-
-      //only construct the type if it was not constructed already
-      if (null == _authorizer)
-      {
-        lock (_syncRoot)
-        {
-          if (null == _authorizer)
-            _authorizer = (IAuthorizeDataPortal)Activator.CreateInstance(authProviderType);
-        }
-      }
-
-      if (InterceptorType != null)
-      {
-        if (_interceptor == null)
-        {
-          lock (_syncRoot)
-          {
-            if (_interceptor == null)
-              _interceptor = (IInterceptDataPortal)Activator.CreateInstance(InterceptorType);
-          }
-        }
-      }
-    }
-
-    private static Type GetAuthProviderType(string cslaAuthorizationProviderAppSettingName)
-    {
-      if (cslaAuthorizationProviderAppSettingName == null)
-        throw new ArgumentNullException("cslaAuthorizationProviderAppSettingName", Resources.AuthorizationProviderNameNotSpecified);
-
-
-      if (null == _authorizer)//not yet instantiated
-      {
-        var authProvider = ConfigurationManager.AppSettings[cslaAuthorizationProviderAppSettingName];
-        return string.IsNullOrEmpty(authProvider) ?
-          typeof(NullAuthorizer) :
-          Type.GetType(authProvider, true);
-
-      }
-      else
-        return _authorizer.GetType();
-
+      Options = options;
+      Authorizer = authorizer;
+      InterceptorList = interceptors;
+      FactoryLoader = factoryLoader;
+      Activator = activator;
+      ExceptionInspector = exceptionInspector;
     }
 
     #region Data Access
```

Comment:
I think we need to think about the order of execution of the latest set of behaviours we have created. I have used an interceptor for triggering revalidation, but interceptors currently run prior to authorization being performed.

This might put the data portal at risk of amplification attacks, where a small action by an unauthenticated user/malicious agent - namely submission of an invalid request - can cause a significant amount of work - namely revalidation of all of the business rules.

I think authorization should come as early as possible - and that means before revalidation.

I can think of three options for solving this:

1. Move the call to AuthorizeRequest into the Initialize method. However, this makes it less visible.
2. Move the call to the interceptors out of the Initialize method and into its own method that is called after AuthorizeRequest.
3. Implement authorization using an interceptor instead.

Option 3 is the most flexible, because it allows people to put an interceptor before authorization if they so desire. However, this is less secure, in that people may not understand the risk of doing so. Generally, security should top flexibility if all else is equal, so option 2 seems the best to me purely from a security perspective.

Reply:
I am trying to remember why the interceptor Initialize method is called before authorization. On the surface it surely seems like authorization should be called before initialize.

I wonder if the thought was that someone might write authorizer code that required the DI container be initialized, and so we called initialize first. I'm thinking that was the reason for the order that exists now.

However, there's no longer a valid scenario for someone to be initializing a DI container, because they should be integrating into the .NET DI subsystem. So the original purpose of the initialize/complete interceptor methods is entirely gone, and has morphed into a new purpose: things like running rules, etc.

So I think the answer is simply to move the `AuthorizeRequest` method above the call to `Initialize`, restoring authorization to the first thing that happens after deserialization.